torch/csrc/jit/passes/inline_forked_closures.cpp

```diff
@@ -58,6 +58,56 @@ void inlineForkedClosure(Node* fork_closure, NodeKind genKind) {
   runCleanupPasses(fork_graph);
 }
 
+void inlineAwaitableThenClosure(Node* then_closure) {
+  Node* function_context_node = then_closure->inputs()[0]->node();
+
+  if (function_context_node->inputs().size() != 2 ||
+      function_context_node->inputs().at(0)->node()->kind() != prim::Closure ||
+      function_context_node->inputs().at(1)->node()->kind() !=
+          prim::TupleConstruct) {
+    throw ErrorReport(then_closure->sourceRange()) << "Cannot fork this value";
+  }
+
+  Node* function = function_context_node->inputs().at(0)->node();
+  Node* context = function_context_node->inputs().at(1)->node();
+  auto then_graph = function->g(attr::Subgraph)->copy();
+  auto g = then_closure->owningGraph();
+  Node* then_node = g->create(prim::awaitable_then, 0)
+                        ->insertAfter(then_closure)
+                        ->setSourceRange(then_closure->sourceRange());
+  Node* then_input_node = g->create(prim::awaitable_then_input, 1)
```

Comment:
I don't really understand why you need this, can you document it here in detail? And maybe add some code pointers in other places where awaitable_the_input shows up that point to the documentation.

Reply:
To add `awaitable_then` to IR - I need to specify the `jit.Value` for its input.

At the moment of `awaitable_then` call - we do not have it, as it will be the result of `await_fn`.
`fork` and `awaitable` do not have this problem as they get real IR arguments at the moment of creation.

```
graph(%x.1 : Tensor):
  %19 : int = prim::Constant[value=1]()
  %7 : Function = prim::Constant[name="gap"]()
  %aw.1 : Await(Tensor) = prim::awaitable_0(%x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:399:17
  %z.1 : Tensor = prim::CallFunction(%7, %x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:400:16
  %40 : Tensor = prim::awaitable_then_input(%aw.1)
   = prim::awaitable_then_1(%aw.1, %40)
  %y.1 : Tensor = prim::awaitable_wait(%aw.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:403:16
  %20 : Tensor = aten::add(%x.1, %y.1, %19) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  %23 : Tensor = aten::add(%20, %z.1, %19) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  return (%23)
with prim::awaitable_0 = graph(%x.1 : Tensor):
  %1 : Function = prim::Constant[name="delayed"]()
  %3 : Tensor = prim::CallFunction(%1, %x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:399:17
  return (%3)
with prim::awaitable_then_1 = graph(%aw.5 : Await(Tensor),
      %aw_out : Tensor):
  %3 : Function = prim::Constant[name="then"]()
  %4 : Tensor = prim::CallFunction(%3, %aw.5, %aw_out) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:401:12
  return (%4)
  ```
  
  If not to specify it, but just add `%aw_out` to await_then subgraph - postprocessing of IR will fail on validation that `%aw_out` gets out of nowhere ( I can not remember the particular error and its location )

Reply:
Added the comment with IR example to `torch/csrc/jit/passes/inline_forked_closures.cpp`